src/compiler/utilitiesPublic.ts

```diff
@@ -1975,67 +1975,87 @@ export function isTemplateLiteral(node: Node): node is TemplateLiteral {
         || kind === SyntaxKind.NoSubstitutionTemplateLiteral;
 }
 
-export function isLeftHandSideExpression(node: Node): node is LeftHandSideExpression {
-    return isLeftHandSideExpressionKind(skipPartiallyEmittedExpressions(node).kind);
-}
+enum NodeKindFacts {
+    None = 0,
+    StrictlyLeftHandSideExpression = 1 << 0,
+    StrictlyUnaryExpression = 1 << 1,
+    StrictlyMiscellaneousExpression = 1 << 2,
+    CanHaveSymbol = 1 << 3,
+    CanHaveLocals = 1 << 4,
+    
+    // TODO: reorganize these
+    Declaration = 1 << 5,
+    DeclarationStatement = 1 << 6,
+    StrictlyStatement = 1 << 7,
+
+    IsLeftHandSideExpression = StrictlyLeftHandSideExpression,
+    IsUnaryExpression = StrictlyUnaryExpression | IsLeftHandSideExpression,
+    IsExpression = StrictlyMiscellaneousExpression | IsUnaryExpression,
+}
+
+const nodeKindFacts: NodeKindFacts[] = [];

```

Comment:
I think you'd actually have better perf filling pre-filling this with numbers to get one of those special compact arrays

Reply:
I'm pretty sure that happens today with `push` according to https://v8.dev/blog/elements-kinds and [this test](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:javascript,selection:(endColumn:15,endLineNumber:7,positionColumn:15,positionLineNumber:7,selectionStartColumn:15,selectionStartLineNumber:7,startColumn:15,startLineNumber:7),source:'let+x+%3D+%5B%5D%3B%0A%0Aconsole.log(%22Unfilled:%22)%3B%0A%25DebugPrint(x)%3B%0A%0Afor+(let+i+%3D+0%3B+i+%3C+10%3B+i%2B%2B)+%7B%0A++++x.push(i)%3B%0A%7D%0A%0Aconsole.log(%22Filled:%22)%3B%0A%25DebugPrint(x)%3B'),l:'5',n:'1',o:'Javascript+source+%231',t:'0')),k:29.092558561075823,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:v8trunk,filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',debugCalls:'1',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'1',trim:'1',verboseDemangling:'0'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:javascript,libs:!(),options:'',overrides:!(),selection:(endColumn:17,endLineNumber:1,positionColumn:17,positionLineNumber:1,selectionStartColumn:17,selectionStartLineNumber:1,startColumn:17,startLineNumber:1),source:1),l:'5',n:'0',o:'+v8+(trunk)+(Editor+%231)',t:'0')),k:19.517197608011134,l:'4',m:100,n:'0',o:'',s:0,t:'0'),(g:!((h:output,i:(compilerName:'.NET+CoreCLR+(main)',editorid:1,fontScale:14,fontUsePx:'0',j:1,wrap:'1'),l:'5',n:'0',o:'Output+of+v8+(trunk)+(Compiler+%231)',t:'0')),header:(),k:51.39024383091303,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4):

```js
let x = [];

console.log("Unfilled:");
%DebugPrint(x);

for (let i = 0; i < 10; i++) {
    x.push(i);
}

console.log("Filled:");
%DebugPrint(x);
```

<details>
<summary>Output</summary>

```
Unfilled:
DebugPrint: 0xcfb0004ab5d: [JSArray]
 - map: 0x0cfb0018e6b9 <Map[16](PACKED_SMI_ELEMENTS)> [FastProperties]
 - prototype: 0x0cfb0018e92d <JSArray[0]>
 - elements: 0x0cfb000006cd <FixedArray[0]> [PACKED_SMI_ELEMENTS]
 - length: 0
 - properties: 0x0cfb000006cd <FixedArray[0]>
 - All own properties (excluding elements): {
    0xcfb00000d41: [String] in ReadOnlySpace: #length: 0x0cfb0028741d <AccessorInfo name= 0x0cfb00000d41 <String[6]: #length>, data= 0x0cfb00000061 <undefined>> (const accessor descriptor), location: descriptor
 }
0xcfb0018e6b9: [Map] in OldSpace
 - map: 0x0cfb00183c35 <MetaMap (0x0cfb00183c85 <NativeContext[285]>)>
 - type: JS_ARRAY_TYPE
 - instance size: 16
 - inobject properties: 0
 - unused property fields: 0
 - elements kind: PACKED_SMI_ELEMENTS
 - enum length: invalid
 - back pointer: 0x0cfb00000061 <undefined>
 - prototype_validity cell: 0x0cfb00000a31 <Cell value= 1>
 - instance descriptors #1: 0x0cfb0018ef45 <DescriptorArray[1]>
 - transitions #1: 0x0cfb0018ef61 <TransitionArray[4]>Transition array #1:
     0x0cfb00000e05 <Symbol: (elements_transition_symbol)>: (transition to HOLEY_SMI_ELEMENTS) -> 0x0cfb0018ef79 <Map[16](HOLEY_SMI_ELEMENTS)>

 - prototype: 0x0cfb0018e92d <JSArray[0]>
 - constructor: 0x0cfb0018e625 <JSFunction Array (sfi = 0xcfb0028c6ad)>
 - dependent code: 0x0cfb000006dd <Other heap object (WEAK_ARRAY_LIST_TYPE)>
 - construction counter: 0

Filled:
DebugPrint: 0xcfb0004ab5d: [JSArray]
 - map: 0x0cfb0018e6b9 <Map[16](PACKED_SMI_ELEMENTS)> [FastProperties]
 - prototype: 0x0cfb0018e92d <JSArray[0]>
 - elements: 0x0cfb0004ab6d <FixedArray[17]> [PACKED_SMI_ELEMENTS]
 - length: 10
 - properties: 0x0cfb000006cd <FixedArray[0]>
 - All own properties (excluding elements): {
    0xcfb00000d41: [String] in ReadOnlySpace: #length: 0x0cfb0028741d <AccessorInfo name= 0x0cfb00000d41 <String[6]: #length>, data= 0x0cfb00000061 <undefined>> (const accessor descriptor), location: descriptor
 }
 - elements: 0x0cfb0004ab6d <FixedArray[17]> {
           0: 0
           1: 1
           2: 2
           3: 3
           4: 4
           5: 5
           6: 6
           7: 7
           8: 8
           9: 9
       10-16: 0x0cfb000006e9 <the_hole_value>
 }
0xcfb0018e6b9: [Map] in OldSpace
 - map: 0x0cfb00183c35 <MetaMap (0x0cfb00183c85 <NativeContext[285]>)>
 - type: JS_ARRAY_TYPE
 - instance size: 16
 - inobject properties: 0
 - unused property fields: 0
 - elements kind: PACKED_SMI_ELEMENTS
 - enum length: invalid
 - back pointer: 0x0cfb00000061 <undefined>
 - prototype_validity cell: 0x0cfb00000a31 <Cell value= 1>
 - instance descriptors #1: 0x0cfb0018ef45 <DescriptorArray[1]>
 - transitions #1: 0x0cfb0018ef61 <TransitionArray[4]>Transition array #1:
     0x0cfb00000e05 <Symbol: (elements_transition_symbol)>: (transition to HOLEY_SMI_ELEMENTS) -> 0x0cfb0018ef79 <Map[16](HOLEY_SMI_ELEMENTS)>

 - prototype: 0x0cfb0018e92d <JSArray[0]>
 - constructor: 0x0cfb0018e625 <JSFunction Array (sfi = 0xcfb0028c6ad)>
 - dependent code: 0x0cfb000006dd <Other heap object (WEAK_ARRAY_LIST_TYPE)>
 - construction counter: 0

Compiler returned: 0
```

</details>